sklearn/impute/_knn.py

```diff
@@ -0,0 +1,271 @@
+import numpy as np
+
+from ..base import BaseEstimator, TransformerMixin
+from ..utils.validation import FLOAT_DTYPES
+from ..metrics import pairwise_distances
+from ..metrics.pairwise import _NAN_METRICS
+from ..neighbors.base import _get_weights
+from ..neighbors.base import _check_weights
+from ..utils import check_array
+from ..utils import is_scalar_nan
+from ..utils.mask import _get_mask
+from ..utils.validation import check_is_fitted
+
+
+class KNNImputer(TransformerMixin, BaseEstimator):
+    """Imputation for completing missing values using k-Nearest Neighbors.
+
+    Each sample's missing values are imputed using values from `n_neighbors`
+    nearest neighbors found in the training set. Each missing feature is then
+    imputed as the average, either weighted or unweighted, of these neighbors.
+    If a sample has more than one feature missing, then the neighbors for that
+    sample can be different depending on the particular feature being imputed.
+    When the number of available neighbors is less than `n_neighbors` and
+    there are no defined distances to the training set, the training set
```

Comment:
"no defined distances" is a bit unclear. And as @jnothman said, I feel this level of detail should be in the user guide.